x/auth/simulation/genesis.go

```diff
@@ -92,35 +92,14 @@ func RandomizedGenState(simState *module.SimulationState) {
 
 // RandomGenesisAccounts returns randomly generated genesis accounts
 func RandomGenesisAccounts(simState *module.SimulationState) (genesisAccs exported.GenesisAccounts) {
-	for i, acc := range simState.Accounts {
+	for _, acc := range simState.Accounts {
 		coins := sdk.Coins{sdk.NewCoin(sdk.DefaultBondDenom, sdk.NewInt(simState.InitialStake))}
 		bacc := types.NewBaseAccountWithAddress(acc.Address)
 		if err := bacc.SetCoins(coins); err != nil {
 			panic(err)
 		}
 
-		var gacc exported.GenesisAccount = &bacc
-
-		// Only consider making a vesting account once the initial bonded validator
-		// set is exhausted due to needing to track DelegatedVesting.
-		if int64(i) > simState.NumBonded && simState.Rand.Intn(100) < 50 {
-			var endTime int64
-
-			startTime := simState.GenTimestamp.Unix()
-
-			// Allow for some vesting accounts to vest very quickly while others very slowly.
-			if simState.Rand.Intn(100) < 50 {
-				endTime = int64(simulation.RandIntBetween(simState.Rand, int(startTime)+1, int(startTime+(60*60*24*30))))
-			} else {
-				endTime = int64(simulation.RandIntBetween(simState.Rand, int(startTime)+1, int(startTime+(60*60*12))))
-			}
-
-			if simState.Rand.Intn(100) < 50 {
-				gacc = types.NewContinuousVestingAccount(&bacc, startTime, endTime)
```

Comment:
undo, there should only be 1 simulation logic per module. It's ok if this file imports the `vestingtypes` package

Reply:
This creates import cycles in `simapp.go`

```
package github.com/cosmos/cosmos-sdk/simapp
        imports github.com/cosmos/cosmos-sdk/x/auth
        imports github.com/cosmos/cosmos-sdk/x/auth/simulation
        imports github.com/cosmos/cosmos-sdk/x/auth/vesting/internal/types
        imports github.com/cosmos/cosmos-sdk/x/auth
```

In general, I don't think the parent module can import its submodule anywhere without creating import cycles.

Reply:
Given the restriction that a module cannot import its submodule, augmenting SimState with an `AccountGenerator` function, as I did in https://github.com/cosmos/cosmos-sdk/pull/5040/commits/2c35f8ce0d974c9f4d5ced138e253e2c652b7863,  does indeed work. 

In `types/module/simulation.go`:
```
type SimulationState struct {
	AppParams        simulation.AppParams
	Cdc              *codec.Codec                                        // application codec
	Rand             *rand.Rand                                          // random number
	GenState         map[string]json.RawMessage                          // genesis state
	Accounts         []simulation.Account                                // simulation accounts
	AccountGenerator func(*SimulationState) authexported.GenesisAccounts // simulation account generator function
	InitialStake     int64                                               // initial coins per account
	NumBonded        int64                                               // number of initially bonded acconts
	GenTimestamp     time.Time                                           // genesis timestamp
	UnbondTime       time.Duration                                       // staking unbond time stored to use it as the slashing maximum evidence duration
}
```

From there, `simapp/state.go` defines `simState` as:

```
simState := &module.SimulationState{
		AppParams:        appParams,
		Cdc:              cdc,
		Rand:             r,
		GenState:         genesisState,
		Accounts:         accs,
		AccountGenerator: vesting.RandomGenesisAccounts,
		InitialStake:     initialStake,
		NumBonded:        numInitiallyBonded,
		GenTimestamp:     genesisTimestamp,
	}
```

And then `/auth/simulation/genesis.go` can call `simState.AccountGenerator(simState)`, which calls `vesting.RandomGenesisAccounts(simState)`.


I tested this using `go test ./simapp -run=TestFullAppSimulation -Enabled=true -v -Commit -Seed=44` and indeed vesting accounts are created in the simulation without import cycles. 

All of this is suggestive to me that the boundaries between a module and its simulation state may not work well in the case of submodules, and could require a rethink. I'd be happy to implement something else if you think there's a better solution.